File: sdk-source/python/aegis_auth.py
"""
AegisAuth — official Python SDK for the Aegis Authentication platform.

Single-file, zero-dependency client (standard library only — no `requests`
required). Drop `aegis_auth.py` into any project:

    from aegis_auth import AegisAuth

    aegis_app = AegisAuth(
        name="My Application",        # display name of your application
        ownerid="YOUR-APPLICATION-KEY",  # Application Key from the dashboard
        secret="",                    # optional API key (elevated calls only)
        version="1.0.0",              # version of the build you ship
    )

    aegis_app.init()
    aegis_app.login("ada", "correct horse battery staple")
    print(aegis_app.user_data.username)

Constructor parameter mapping:
    name     -> application display name (informational)
    ownerid  -> Application Key (public key), sent as the `x-app-key` header
    secret   -> optional API key for elevated calls, sent as `x-api-key`
    version  -> client version, checked against the minimum published version
    url      -> optional custom API base URL (self-hosted deployments)

Works on CPython 3.8+ on Windows, macOS and Linux.
"""

from __future__ import annotations

import getpass
import hashlib
import json
import platform
import subprocess
import time
import urllib.error
import urllib.request
import uuid
from datetime import datetime, timezone

__version__ = "1.0.0"
__all__ = ["AegisAuth", "DEFAULT_BASE_URL"]

DEFAULT_BASE_URL = "https://project--9347818a-431f-4584-98ac-b0d367707e9b.lovable.app"


# ---------------------------------------------------------------------------
# Data containers
# ---------------------------------------------------------------------------

class _Response:
    """Outcome of the most recent call. Check `response.success` after every method."""

    def __init__(self) -> None:
        self.success: bool = False
        self.message: str = ""


class Subscription:
    """An active license attached to the signed-in user (or key-only session)."""

    def __init__(self, key: str, status: str, expires_at: str | None) -> None:
        self.subscription = key
        self.status = status
        self.expiry = _parse_date(expires_at)

    @property
    def timeleft(self):
        if not self.expiry:
            return None
        return self.expiry - datetime.now(timezone.utc)


class _UserData:
    """Data about the signed-in user. Populated by login/register/license/check."""

    def __init__(self) -> None:
        self.username: str | None = None
        self.email: str | None = None
        self.status: str | None = None
        self.hwid: str | None = None
        self.createdate: datetime | None = None
        self.lastlogin: datetime | None = None
        self.logincount: int = 0
        self.subscriptions: list[Subscription] = []

    def clear(self) -> None:
        self.__init__()


class _AppData:
    """Application-wide data. Populated by init/fetchstats."""

    def __init__(self) -> None:
        self.numUsers: int = 0
        self.numKeys: int = 0
        self.app_ver: str = ""
        self.maintenance: bool = False
        self.maintenance_message: str | None = None
        self.hwid_required: bool = False
        self.session_timeout_minutes: int = 0
        self.server_time: str | None = None
        self.downloads: list[dict] = []


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------

class AegisAuth:
    """AegisAuth client. Create one instance per application."""

    def __init__(
        self,
        name: str,
        ownerid: str,
        secret: str,
        version: str,
        url: str = DEFAULT_BASE_URL,
    ) -> None:
        if not name:
            raise ValueError("name is required")
        if not ownerid:
            raise ValueError("ownerid (Application Key) is required")

        self.name = name
        self.ownerid = ownerid
        self.secret = secret or ""
        self.version = version or "1.0.0"
        self._base_url = url.rstrip("/") + "/api/public/v1/"

        self.sessionid: str | None = None
        self.hwid: str = _hardware_id()

        self.user_data = _UserData()
        self.app_data = _AppData()
        self.response = _Response()

    # ------------------------------------------------------------------
    # Core
    # ------------------------------------------------------------------

    def init(self) -> None:
        """Handshake with the server. Call once before anything else; fills app_data."""
        data = self._req("init", {"version": self.version})
        if not self.response.success or not data:
            return

        version_info = data.get("version") or {}
        self.app_data.app_ver = version_info.get("current", self.version)
        self.app_data.maintenance = bool(data.get("maintenance"))
        self.app_data.maintenance_message = data.get("maintenance_message")
        self.app_data.hwid_required = bool(data.get("hwid_required"))
        self.app_data.session_timeout_minutes = int(data.get("session_timeout_minutes") or 0)
        self.app_data.server_time = data.get("server_time")

        if version_info.get("update_required"):
            self._error("An update is required before you can use this build.")
            return
        self.response.message = f"Initialized {self.name}"

    def register(self, user: str, password: str, license_key: str | None = None,
                 email: str | None = None) -> None:
        """Register a new user, optionally redeeming a license key."""
        body = {"username": user, "password": password, "hwid": self.hwid}
        if license_key:
            body["license_key"] = license_key
        if email:
            body["email"] = email
        data = self._req("register", body)
        if self.response.success:
            self._store_auth(data, "Registered successfully")

    def login(self, user: str, password: str) -> None:
        """Authenticate a user and open a session."""
        data = self._req("login", {"username": user, "password": password, "hwid": self.hwid})
        if self.response.success:
            self._store_auth(data, "Logged in successfully")

    def license(self, key: str) -> None:
        """Key-only authentication: validates a license key and binds it to this machine."""
        check = self._req("license/validate", {"license_key": key, "hwid": self.hwid})
        if not self.response.success:
            return
        activation = self._req("license/activate", {"license_key": key, "hwid": self.hwid})
        if not self.response.success:
            return

        lic = (activation or {}).get("license") or (check or {}).get("license")
        self.user_data.username = None
        self.user_data.hwid = self.hwid
        self.user_data.subscriptions = [_subscription(lic)] if lic else []
        self.response.message = "License activated"

    def upgrade(self, user: str, key: str) -> None:
        """Attach a license key to an existing user account."""
        self._req("license/activate", {
            "license_key": key,
            "hwid": self.hwid,
            "username": user,
        })
        if self.response.success:
            self.response.message = f"License attached to {user}"

    def var(self, varid: str) -> str | None:
        """Read an application variable published from the dashboard."""
        data = self._req("variables/get", {"scope": "application"})
        if not self.response.success or not data:
            return None
        variables = data.get("variables") or {}
        if varid in variables:
            return variables[varid]
        self._error(f"Variable not found: {varid}")
        return None

    def getvar(self, varname: str) -> str | None:
        """Read a per-user variable (requires an active session)."""
        data = self._req("variables/get", {"scope": "user"})
        if not self.response.success or not data:
            return None
        return (data.get("variables") or {}).get(varname)

    def setvar(self, varname: str, value: str) -> None:
        """Write a per-user variable (requires an active session)."""
        self._req("variables/set", {"scope": "user", "key": varname, "value": value or ""})
        if self.response.success:
            self.response.message = "Variable saved"

    def log(self, message: str) -> None:
        """Write a message to the application's audit log on the dashboard."""
        self._req("log", {
            "message": message or "",
            "pcuser": getpass.getuser(),
            "hwid": self.hwid,
        })
        if self.response.success:
            self.response.message = "Logged"

    def check(self) -> bool:
        """Validate the active session against the server. Fills user_data."""
        if not self.sessionid:
            self._error("No active session")
            return False
        data = self._req("session/check", {})
        if not self.response.success or not data or not data.get("valid"):
            return False
        self._fill_user(data.get("user"), data.get("license"))
        return True

    def logout(self) -> None:
        """Terminate the active session on the server."""
        self._req("logout", {})
        self.sessionid = None
        self.user_data.clear()
        if self.response.success:
            self.response.message = "Logged out"

    def fetchstats(self) -> None:
        """Refresh app_data with live counters, versions and download links."""
        data = self._req("app/data", {})
        if not self.response.success or not data:
            return
        stats = data.get("stats") or {}
        self.app_data.numUsers = int(stats.get("users") or 0)
        self.app_data.numKeys = int(stats.get("licenses") or 0)
        application = data.get("application") or {}
        self.app_data.app_ver = application.get("current_version", self.app_data.app_ver)
        self.app_data.downloads = [
            {"name": d.get("name"), "url": d.get("file_url")}
            for d in data.get("downloads") or []
        ]
        self.response.message = "Stats fetched"

    def update_available(self, channel: str = "stable") -> bool:
        """True when a newer build than `version` has been published."""
        data = self._req("version/check", {"version": self.version, "channel": channel})
        return bool(self.response.success and data and data.get("update_available"))

    def use_session(self, token: str) -> None:
        """Restore a session token persisted earlier (skip the login screen)."""
        self.sessionid = token

    def expirydaysleft(self) -> int:
        """Days remaining on the active license, or 0 when none/expired."""
        if not self.user_data.subscriptions:
            return 0
        expiry = self.user_data.subscriptions[0].expiry
        if not expiry:
            return 0
        return max(0, (expiry - datetime.now(timezone.utc)).days)

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _store_auth(self, data: dict, message: str) -> None:
        session = (data or {}).get("session") or {}
        if session.get("token"):
            self.sessionid = session["token"]
        self._fill_user((data or {}).get("user"), (data or {}).get("license"))
        self.response.message = message

    def _fill_user(self, user: dict | None, license_: dict | None) -> None:
        if user:
            self.user_data.username = user.get("username")
            self.user_data.email = user.get("email")
            self.user_data.status = user.get("status")
            self.user_data.hwid = user.get("hwid")
            self.user_data.createdate = _parse_date(user.get("created_at"))
            self.user_data.lastlogin = _parse_date(user.get("last_login_at"))
            self.user_data.logincount = int(user.get("login_count") or 0)
        self.user_data.subscriptions = [_subscription(license_)] if license_ else []

    def _error(self, message: str) -> None:
        self.response.success = False
        self.response.message = message

    def _req(self, endpoint: str, body: dict) -> dict | None:
        """POST to an endpoint, normalise the envelope, fill `response`."""
        headers = {
            "content-type": "application/json",
            "x-app-key": self.ownerid,
            "x-timestamp": str(int(time.time())),
            "user-agent": f"aegisauth-python/{__version__}",
        }
        if self.secret:
            headers["x-api-key"] = self.secret
        if self.sessionid:
            headers["x-session-token"] = self.sessionid

        request = urllib.request.Request(
            self._base_url + endpoint,
            data=json.dumps(body).encode("utf-8"),
            headers=headers,
            method="POST",
        )

        try:
            with urllib.request.urlopen(request, timeout=30) as handle:
                payload = json.loads(handle.read().decode("utf-8") or "{}")
        except urllib.error.HTTPError as err:
            try:
                payload = json.loads(err.read().decode("utf-8") or "{}")
            except Exception:
                self._error(f"Network error: HTTP {err.code}")
                return None
        except Exception as err:  # URLError, timeout, JSON decode, ...
            self._error(f"Network error: {err}")
            return None

        if not payload.get("success"):
            error = payload.get("error") or {}
            self._error(error.get("message") or "Request failed.")
            return None

        self.response.success = True
        self.response.message = "Success"
        return payload.get("data") or {}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _subscription(license_: dict | None) -> Subscription:
    license_ = license_ or {}
    return Subscription(license_.get("key"), license_.get("status"), license_.get("expires_at"))


def _parse_date(iso: str | None) -> datetime | None:
    if not iso:
        return None
    try:
        parsed = datetime.fromisoformat(iso.replace("Z", "+00:00"))
        return parsed if parsed.tzinfo else parsed.replace(tzinfo=timezone.utc)
    except ValueError:
        return None


def _hardware_id() -> str:
    """Stable per-machine identifier (SHA-256 hashed; raw machine id never leaves the device)."""
    system = platform.system()
    seed = ""
    try:
        if system == "Windows":
            seed = subprocess.check_output(
                "wmic csproduct get uuid", shell=True, timeout=5
            ).decode(errors="ignore").split("\n")[1].strip()
        elif system == "Darwin":
            seed = subprocess.check_output(
                "ioreg -rd1 -c IOPlatformExpertDevice | grep IOPlatformUUID",
                shell=True, timeout=5,
            ).decode(errors="ignore").split('"')[-2]
        else:
            with open("/etc/machine-id", "r", encoding="utf-8") as handle:
                seed = handle.read().strip()
    except Exception:
        seed = ""
    if not seed:
        seed = f"{platform.node()}|{getpass.getuser()}|{uuid.getnode()}"
    return hashlib.sha256(f"aegisauth:{seed}".encode()).hexdigest()

File: sdk-source/cpp/src/AegisAuth.cpp
// AegisAuth C++ SDK — implementation. See include/AegisAuth/AegisAuth.hpp.
#include "AegisAuth/AegisAuth.hpp"

#include <array>
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <ctime>
#include <sstream>
#include <thread>

#include <curl/curl.h>

#if defined(_WIN32)
#include <windows.h>
#else
#include <unistd.h>
#endif

namespace {

size_t writeCallback(void* contents, size_t size, size_t nmemb, std::string* out) {
    out->append(static_cast<const char*>(contents), size * nmemb);
    return size * nmemb;
}

std::string trimSlashes(std::string value) {
    while (!value.empty() && value.back() == '/') value.pop_back();
    return value;
}

/** FNV-1a based hex digest — stable across runs, never reveals raw machine facts. */
std::string digest(const std::string& input) {
    uint64_t h = 1469598103934665603ULL;
    for (unsigned char c : input) {
        h ^= c;
        h *= 1099511628211ULL;
    }
    std::ostringstream out;
    for (int round = 0; round < 4; ++round) {
        h ^= h >> 33;
        h *= 0xff51afd7ed558ccdULL;
        out << std::hex << h;
    }
    return out.str().substr(0, 64);
}

/** Days between an ISO-8601 UTC timestamp ("2026-02-01T00:00:00Z") and now. */
int daysUntil(const std::string& iso) {
    if (iso.size() < 19) return 0;
    std::tm tm{};
    if (sscanf(iso.c_str(), "%d-%d-%dT%d:%d:%d", &tm.tm_year, &tm.tm_mon, &tm.tm_mday,
               &tm.tm_hour, &tm.tm_min, &tm.tm_sec) != 6)
        return 0;
    tm.tm_year -= 1900;
    tm.tm_mon -= 1;
#if defined(_WIN32)
    std::time_t expiry = _mkgmtime(&tm);
#else
    std::time_t expiry = timegm(&tm);
#endif
    if (expiry < 0) return 0;
    const double days = std::difftime(expiry, std::time(nullptr)) / 86400.0;
    return days > 0 ? static_cast<int>(days + 0.999) : 0;
}

}  // namespace

namespace AegisAuth {

using aegisauth::Json;
using aegisauth::JsonObject;

api::api(std::string name, std::string ownerid, std::string secret, std::string version,
         std::string url)
    : name(std::move(name)),
      ownerid(std::move(ownerid)),
      secret(std::move(secret)),
      version(std::move(version)),
      url(trimSlashes(std::move(url))) {
    if (this->ownerid.empty()) throw std::invalid_argument("ownerid (Application Key) is required");
    hwid = hardware_id();
    curl_global_init(CURL_GLOBAL_DEFAULT);
}

std::string api::hardware_id() {
    std::string facts;
#if defined(_WIN32)
    char name[256] = {0};
    DWORD size = sizeof(name);
    GetComputerNameA(name, &size);
    facts += name;
    if (const char* user = std::getenv("USERNAME")) facts += user;
    facts += "windows";
#else
    std::array<char, 256> host{};
    if (gethostname(host.data(), host.size()) == 0) facts += host.data();
    if (const char* user = std::getenv("USER")) facts += user;
    facts += "posix";
#endif
    return digest(facts);
}

Json api::req(const std::string& endpoint, const JsonObject& body) {
    const std::string fullUrl = url + "/api/public/v1/" + endpoint;
    const std::string payload = Json(body).dump();
    std::string lastNetworkError;

    for (int attempt = 0; attempt <= 2; ++attempt) {
        CURL* curl = curl_easy_init();
        if (!curl) {
            error("Unable to create an HTTP client.");
            return Json();
        }

        curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "content-type: application/json");
        headers = curl_slist_append(headers, "user-agent: aegisauth-cpp/1.0.0");
        headers = curl_slist_append(headers, ("x-app-key: " + ownerid).c_str());
        headers = curl_slist_append(
            headers, ("x-timestamp: " + std::to_string(std::time(nullptr))).c_str());
        if (!secret.empty()) headers = curl_slist_append(headers, ("x-api-key: " + secret).c_str());
        if (!sessionid.empty())
            headers = curl_slist_append(headers, ("x-session-token: " + sessionid).c_str());

        std::string text;
        curl_easy_setopt(curl, CURLOPT_URL, fullUrl.c_str());
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &text);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);

        CURLcode code = curl_easy_perform(curl);
        long status = 0;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status);
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);

        if (code != CURLE_OK) {
            lastNetworkError = curl_easy_strerror(code);
            std::this_thread::sleep_for(std::chrono::milliseconds(250 * (attempt + 1)));
            continue;
        }
        if (status >= 500 && attempt < 2) {
            std::this_thread::sleep_for(std::chrono::milliseconds(250 * (attempt + 1)));
            continue;
        }

        Json envelope = Json::parse(text.empty() ? "{}" : text);
        if (!envelope["success"].asBool()) {
            error(envelope["error"]["message"].asString("Request failed."));
            return Json();
        }
        response.success = true;
        response.message = "Success";
        return envelope["data"];
    }

    error("Network error: " + (lastNetworkError.empty() ? "request failed" : lastNetworkError));
    return Json();
}

void api::error(std::string message) {
    response.success = false;
    response.message = std::move(message);
}

void api::store_auth(const Json& data, const char* message) {
    const std::string token = data["session"]["token"].asString();
    if (!token.empty()) sessionid = token;
    fill_user(data["user"], data["license"]);
    response.message = message;
}

void api::fill_user(const Json& user, const Json& license) {
    user_data.username = user["username"].asString();
    user_data.email = user["email"].asString();
    user_data.status = user["status"].asString();
    user_data.hwid = user["hwid"].asString();
    user_data.createdate = user["created_at"].asString();
    user_data.lastlogin = user["last_login_at"].asString();
    user_data.logincount = static_cast<int>(user["login_count"].asNumber());
    user_data.subscriptions.clear();
    const std::string key = license["key"].asString();
    if (!key.empty())
        user_data.subscriptions.push_back(
            {key, license["status"].asString(), license["expires_at"].asString()});
}

void api::init() {
    Json data = req("init", {{"version", Json(version)}});
    if (!response.success) return;

    app_data.app_ver = data["version"]["current"].asString(version);
    app_data.maintenance = data["maintenance"].asBool();
    app_data.maintenance_message = data["maintenance_message"].asString();
    app_data.hwid_required = data["hwid_required"].asBool();
    app_data.session_timeout_minutes =
        static_cast<int>(data["session_timeout_minutes"].asNumber());
    app_data.server_time = data["server_time"].asString();

    if (data["version"]["update_required"].asBool()) {
        error("An update is required before you can use this build.");
        return;
    }
    response.message = "Initialized " + name;
}

void api::login(std::string username, std::string password) {
    Json data = req("login", {{"username", Json(std::move(username))},
                              {"password", Json(std::move(password))},
                              {"hwid", Json(hwid)}});
    if (response.success) store_auth(data, "Logged in successfully");
}

void api::regstr(std::string username, std::string password, std::string key, std::string email) {
    JsonObject body{{"username", Json(std::move(username))},
                    {"password", Json(std::move(password))},
                    {"hwid", Json(hwid)}};
    if (!key.empty()) body.emplace("license_key", Json(std::move(key)));
    if (!email.empty()) body.emplace("email", Json(std::move(email)));
    Json data = req("register", body);
    if (response.success) store_auth(data, "Registered successfully");
}

void api::license(std::string key) {
    Json check = req("license/validate", {{"license_key", Json(key)}, {"hwid", Json(hwid)}});
    if (!response.success) return;
    Json activation = req("license/activate", {{"license_key", Json(key)}, {"hwid", Json(hwid)}});
    if (!response.success) return;

    const Json lic = activation["license"]["key"].asString().empty() ? check["license"]
                                                                     : activation["license"];
    user_data.username.clear();
    user_data.hwid = hwid;
    user_data.subscriptions.clear();
    if (!lic["key"].asString().empty())
        user_data.subscriptions.push_back(
            {lic["key"].asString(), lic["status"].asString(), lic["expires_at"].asString()});
    response.message = "License activated";
}

void api::upgrade(std::string username, std::string key) {
    req("license/activate", {{"license_key", Json(std::move(key))},
                             {"hwid", Json(hwid)},
                             {"username", Json(username)}});
    if (response.success) response.message = "License attached to " + username;
}

std::string api::var(std::string varid) {
    Json data = req("variables/get", {{"scope", Json("application")}});
    if (!response.success) return "";
    const std::string value = data["variables"][varid].asString();
    if (value.empty()) {
        error("Variable not found: " + varid);
        return "";
    }
    return value;
}

std::string api::getvar(std::string varname) {
    Json data = req("variables/get", {{"scope", Json("user")}});
    if (!response.success) return "";
    return data["variables"][varname].asString();
}

void api::setvar(std::string varname, std::string value) {
    req("variables/set", {{"scope", Json("user")},
                          {"key", Json(std::move(varname))},
                          {"value", Json(std::move(value))}});
    if (response.success) response.message = "Variable saved";
}

void api::log(std::string msg) {
    const char* user = std::getenv(
#if defined(_WIN32)
        "USERNAME"
#else
        "USER"
#endif
    );
    req("log", {{"message", Json(std::move(msg))},
                {"pcuser", Json(user ? user : "")},
                {"hwid", Json(hwid)}});
    if (response.success) response.message = "Logged";
}

bool api::check() {
    if (sessionid.empty()) {
        error("No active session");
        return false;
    }
    Json data = req("session/check", {});
    if (!response.success || !data["valid"].asBool()) return false;
    fill_user(data["user"], data["license"]);
    return true;
}

void api::logout() {
    req("logout", {});
    sessionid.clear();
    user_data = user_data_class{};
    if (response.success) response.message = "Logged out";
}

void api::fetchstats() {
    Json data = req("app/data", {});
    if (!response.success) return;
    app_data.numUsers = static_cast<int>(data["stats"]["users"].asNumber());
    app_data.numKeys = static_cast<int>(data["stats"]["licenses"].asNumber());
    app_data.app_ver = data["application"]["current_version"].asString(app_data.app_ver);
    app_data.downloads.clear();
    for (const Json& d : data["downloads"].items())
        app_data.downloads.emplace_back(d["name"].asString(), d["file_url"].asString());
    response.message = "Stats fetched";
}

bool api::update_available(std::string channel) {
    Json data = req("version/check", {{"version", Json(version)},
                                      {"channel", Json(std::move(channel))}});
    return response.success && data["update_available"].asBool();
}

void api::use_session(std::string token) { sessionid = std::move(token); }

int api::expirydaysleft() const {
    if (user_data.subscriptions.empty()) return 0;
    return daysUntil(user_data.subscriptions.front().expiry);
}

}  // namespace AegisAuth